include/alpaka/pltf/PltfGenericSycl.hpp

```diff
@@ -18,89 +18,150 @@
 
 #    include <CL/sycl.hpp>
 
-#    include <iostream>
+#    include <cstddef>
+#    include <exception>
+#    include <functional>
+#    if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+#        include <iostream>
+#    endif
+#    include <mutex>
+#    include <optional>
+#    include <shared_mutex>
 #    include <sstream>
 #    include <stdexcept>
 #    include <vector>
 
 namespace alpaka::experimental
 {
     //! The SYCL device manager.
-    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl>
+    template<typename TSelector>
+    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl<TSelector>>
     {
     public:
         PltfGenericSycl() = delete;
+
+        [[nodiscard]] static auto syclPlatform() -> sycl::platform&
+        {
+            auto lock = std::unique_lock<std::mutex>{platform_mutex};
+
+            if(!platform_opt.has_value())
+                platform_opt = std::make_optional<sycl::platform>(TSelector{});
+
+            return platform_opt.value();
+        }
+
+        [[nodiscard]] static auto syclDevices() -> std::vector<sycl::device>&
+        {
+            auto lock = std::unique_lock<std::mutex>{devices_mutex};
+
+            if(!devices_opt.has_value())
+                devices_opt = std::make_optional<std::vector<sycl::device>>(syclPlatform().get_devices());
+
+            return devices_opt.value();
+        }
+
+        [[nodiscard]] static auto syclContext() -> sycl::context&
+        {
+            auto lock = std::unique_lock<std::mutex>{context_mutex};
+
+            if(!context_opt.has_value())
+            {
+                context_opt = std::make_optional<sycl::context>(
+                    syclDevices(),
+                    [](sycl::exception_list exceptions)
+                    {
+                        auto ss_err = std::stringstream{};
+                        ss_err << "Caught asynchronous SYCL exception(s):\n";
+                        for(std::exception_ptr e : exceptions)
+                        {
+                            try
+                            {
+                                std::rethrow_exception(e);
+                            }
+                            catch(sycl::exception const& err)
+                            {
+                                ss_err << err.what() << " (" << err.code() << ")\n";
+                            }
+                        }
+                        throw std::runtime_error(ss_err.str());
+                    });
+            }
+
+            return context_opt.value();
+        }
+
+        // Use destructor attribute to make sure our static variables are destroyed in the correct order at shutdown.
+        [[gnu::destructor]] static auto reset() -> void
+        {
+            auto lock
+                = std::scoped_lock<std::mutex, std::mutex, std::mutex>{platform_mutex, devices_mutex, context_mutex};
```

Comment:
I guess you also got a clang warning here? I wonder whether we should disable that one. I would prefer:
```suggestion
            auto lock
                = std::scoped_lock{platform_mutex, devices_mutex, context_mutex};
```

Reply:
Yes, that CTAD warning is really annoying. One would think they would guard against the standard library.